src/codec.cairo

```diff
@@ -25,31 +41,57 @@ pub impl EncodeSpan<T, +Encode<T>> of Encode<Span<T>> {
             item.encode_to(ref dest);
         }
     }
+
+    fn weight_to(self: @Span<T>, ref weight: usize, factor: usize) {
+        let items = *self;
+        encode_compact_size_weight(items.len(), ref weight, factor);
+        for item in items {
+            item.weight_to(ref weight, factor);
+        };
+    }
 }
 
 pub impl EncodeByteArray of Encode<ByteArray> {
     fn encode_to(self: @ByteArray, ref dest: ByteArray) {
         encode_compact_size(self.len(), ref dest);
         dest.append(self);
     }
+
+    fn weight_to(self: @ByteArray, ref weight: usize, factor: usize) {
+        let len: usize = self.len();
+        encode_compact_size_weight(len, ref weight, factor);
+        weight += len * factor;
+    }
 }
 
 pub impl EncodeU32 of Encode<u32> {
     fn encode_to(self: @u32, ref dest: ByteArray) {
         dest.append_word_rev((*self).into(), 4);
     }
+
+    fn weight_to(self: @u32, ref weight: usize, factor: usize) {
+        weight += 4 * factor;
+    }
 }
 
 pub impl EncodeU64 of Encode<u64> {
     fn encode_to(self: @u64, ref dest: ByteArray) {
         dest.append_word_rev((*self).into(), 8);
     }
+
+    fn weight_to(self: @u64, ref weight: usize, factor: usize) {
+        weight += 8 * factor;
```

Comment:
It's better to multiple by factors on the top level only (in the transaction encoding)